File: tests/WeavePort.ReuseTests/worker.mjs
import { open, unlink, access } from 'node:fs/promises';
import { join } from 'node:path';
import { randomUUID } from 'node:crypto';
const { PluginApplication } = await import(process.argv[2]);
const app = new PluginApplication();
let previous, oldFile, hidden;
const cache = new Map();
async function exists(path) {
    if (!path) {
        return false;
    }
    try {
        await access(path);
        return true;
    } catch (error) {
        if (error.code === 'ENOENT') {
            return false;
        }
        throw error;
    }
}

app.function('session', async (_, context) => {
    let expired = !previous;
    if (previous) {
        try {
            await previous.callHost('who', {});
        } catch {
            expired = true;
        }
    }
    const fileExists = await exists(oldFile);
    if (cache.size || fileExists) {
        throw new Error('Registered residue');
    }
    const secret = context.configuration.secret;
    cache.set('secret', secret);
    context.onClose(() => cache.clear());
    const path = join(process.env.WEAVEPORT_WORKSPACE || '/tmp', randomUUID() + '.session');
    oldFile = path;
    context.onClose(() => unlink(path));
    const file = context.own(await open(path, 'wx'));
    await file.write(secret);
    previous = context;
    const owner = await context.callHost('who', { tenant: 'forged' });
    return { tenant: context.tenant, secret, expired, callback: owner };
});
app.function('stash', async (_, context) => {
    hidden = context.tenant;
    return { stored: true };
});
app.function('probe', async () => ({ hidden: hidden ?? null }));
app.function('cleanup-fail', async (_, context) => {
    context.onClose(() => {
        throw new Error('Deliberate cleanup failure');
    });
    return {};
});
app.function('cleanup-hang', async (_, context) => {
    context.onClose(() => new Promise((resolve) => setTimeout(resolve, 30000)));
    return {};
});
app.function('delay', async (value, context) => {
    await new Promise((resolve) => setTimeout(resolve, value.ms));
    return { tenant: context.tenant };
});
app.stream('rows', async function* (_, context) {
    previous = context;
    for (let i = 0; i < 20; i++) {
        yield { tenant: context.tenant, i };
    }
});
await app.run();

File: tests/WeavePort.ReuseTests/worker.py
import asyncio
import os
from pathlib import Path
import sys
import uuid

if len(sys.argv) > 1 and sys.argv[1] != "-":
    sys.path.insert(0, sys.argv[1])
from weaveport_sdk import PluginApplication

app = PluginApplication()
previous = None
cache = {}
old_file = None
hidden = None


@app.function("session")
async def session(value, context):
    global previous, old_file
    expired = previous is None
    if previous is not None:
        try:
            await previous.call_host("who", {})
        except RuntimeError:
            expired = True
    if cache or old_file and old_file.exists():
        raise RuntimeError("Registered residue")
    secret = context.configuration["secret"]
    cache["secret"] = secret
    context.on_close(cache.clear)
    old_file = Path(os.getenv("WEAVEPORT_WORKSPACE", "/tmp")) / (
        uuid.uuid4().hex + ".session"
    )
    path = old_file
    context.on_close(lambda: path.unlink())
    stream = context.own(path.open("w"))
    stream.write(secret)
    previous = context
    owner = await context.call_host("who", {"tenant": "forged"})
    return dict(tenant=context.tenant, secret=secret, expired=expired, callback=owner)


@app.function("context")
async def density(value, context):
    local = bytearray(context.configuration["data"], "utf-8")
    context.on_close(local.clear)
    return dict(tenant=context.tenant, configuration=dict(context.configuration))


@app.function("stash")
async def stash(value, context):
    global hidden
    hidden = context.tenant
    return dict(stored=True)


@app.function("probe")
async def probe(value, context):
    return dict(hidden=hidden)


@app.function("cleanup-fail")
async def fail(value, context):
    def cleanup():
        raise RuntimeError("Deliberate cleanup failure")

    context.on_close(cleanup)
    return {}


@app.function("cleanup-hang")
async def hang(value, context):
    context.on_close(lambda: asyncio.sleep(30))
    return {}


@app.function("delay")
async def delay(value, context):
    await asyncio.sleep(value["ms"] / 1000)
    return dict(tenant=context.tenant)


@app.stream("rows")
async def rows(value, context):
    global previous
    previous = context
    for i in range(20):
        yield dict(tenant=context.tenant, i=i)


app.run()
